python/lib/scheduler/dmod/scheduler/resources/resource_manager.py

```diff
@@ -170,126 +184,159 @@ def allocate_single_node(self, cpus: int, memory: int) -> List[Optional[Resource
+            Importantly, this is slightly different not just those within
 
-            TODO this is a balanced round robin algorithm, assuming an even distribution is possible across all resources,
-            with up to num_node-1 remainders to fill in.  This is not the most generic "round robin" in which we allocate
+            This is a balanced round robin algorithm, assuming an even distribution is possible across all resources,
+            with up to num_node-1 remainders to fill in.
+            Note that this is not the most generic "round robin" in which we allocate
             cpus one after the other across all available resources and don't try to balance.
             i.e. a request for 10 cpus with an available resource view of [4, 2, 4] would fail to allocate with this
-            algorithm, because it assumes an availablity of [4, 3, 3]
+            algorithm, because it assumes an availability of [4, 3, 3]
+
+            For a ``BUNDLED`` :class:`AllocationAssetGrouping`, a single allocation is returned per :class:`Resource`
+            node containing all the assets allocated from that node. For ``SILO``, several per-node allocations are
+            returned, each with a single CPU and a roughly even share of the total requested memory.
 
             Parameters
             ----------
                 cpus: Total number of CPUs requested
                 memory: Amount of memory required in bytes
+                asset_grouping: The way compute assets from a single node are grouped into allocations
 
             Returns
             -------
             [ResourceAlloction]
                 List of one or more ResourceAllocation if allocation successful, otherwise, [None]
         """
         #TODO consider scaling memory per cpu
-        #Find the number of cpus to allocate to each node
         self.validate_allocation_parameters(cpus, memory)
-        resources = list(self.get_useable_resources())
-        per_alloc_mem = int(memory / cpus)
-
-        num_node = len(resources)
-        if num_node == 0:
+        # Get all active and ready nodes, regardless of whether they are "allocateable" (i.e., not full)
+        resource_nodes = {r.resource_id: r for r in self.get_resources() if r.is_active() and r.is_ready()}
+
+        # Calculate in advance the exact amounts of CPUs and memory per node for the necessary balance
+        # This is slightly different from simply an even share due to discrete amounts and remainders
+        cpus_per_node, memory_per_node = dict(), dict()
+        num_nodes = len(resource_nodes)
+        cpu_share, mem_share = int(cpus / num_nodes), int(memory / num_nodes)
+        cpu_remainder, mem_remainder = cpus - cpu_share, memory - mem_share
+
+        for node_id, node in resource_nodes.items():
+            # We must plan to request at least the per-node amounts on each node
+            if node.cpu_count < cpu_share or node.memory < memory:
```

Comment:
You are correct, but for some reason that was the intent.  I don't remember exactly why; this goes back to @hellkite500's original implementation.   

It may be the case that, if a job can tolerate an unbalanced round robin allocation, it can just as easily function with a `FILL_NODES` allocation.

Reply:
Yeah, in looking at the git blame I have a better understanding. Correct me if im wrong, but it looks like the current [implementation](https://github.com/NOAA-OWP/DMOD/blob/8a4c45e10c7e31abc6620e32414aae702c425bab/python/lib/scheduler/dmod/scheduler/resources/resource_manager.py#L274-L288) is a general round robin, but as you mentioned the [original](https://github.com/NOAA-OWP/DMOD/blame/4a7e464dc3d96edc91bacfc803dc4ffbbbc9455a/python/lib/scheduler/dmod/scheduler/resources/resource_manager.py#L264-L279) algorithm was a node balanced round robin? I don't want to hold this up for this, just trying to get clarification more than anything.

Reply:
Hmm ... again, you are correct, and that appears to have been a bug.  When I changed the implementation to go to 1:1 CPU-to-allocation, I unintentionally altered the behavior of the function beyond just that.  Or so it would seem, given the docstring and the related commit message.  So let's go with balanced as being the expected implementation for now.

Reply:
Sounds good to me!